{
    "active_lock_reason": null,
    "assignee": null,
    "assignees": [],
    "author_association": "CONTRIBUTOR",
    "body": "### Neovim version (nvim -v)\r\n\r\n0.7.0\r\n\r\n### Vim (not Nvim) behaves the same?\r\n\r\nN/A\r\n\r\n### Operating system/version\r\n\r\nCentOS 7.9\r\n\r\n### Terminal name/version\r\n\r\nxfce4-terminal 0.8.7.4\r\n\r\n### $TERM environment variable\r\n\r\ntmux-256color\r\n\r\n### Installation\r\n\r\nsystem package manager\r\n\r\n### How to reproduce the issue\r\n\r\nIn Neovim 0.7.0, I am running a test that does the following steps:\r\n\r\n1. Ask Neovim to send us an rpcrequest for rpc-1\r\n2. Ask Neovim to send us an rpcrequest for rpc-2\r\n3. Wait for those 2 requests to arrive from Neovim\r\n4. Respond to the second request (the one for rpc-2)\r\n5. Respond to the first request (the one for rpc-1)\r\n6. Wait for the 2 responses to steps 1 and 2\r\n\r\nSometimes this succeeds and sometimes it fails. On success, the interaction looks like this:\r\n```\r\nSent: [ 0, 23, \"nvim_call_function\", [ \"rpcrequest\", [ 10, \"rpc-1\" ] ] ]\r\nSent: [ 0, 24, \"nvim_call_function\", [ \"rpcrequest\", [ 10, \"rpc-2\" ] ] ]\r\nReceived: [ 0, 1, \"rpc-1\", [] ]\r\nReceived: [ 0, 2, \"rpc-2\", [] ]\r\nSent: [ 1, 2, nil, nil ]\r\nSent: [ 1, 1, nil, nil ]\r\nReceived: [ 1, 24, nil, nil ]\r\nReceived: [ 1, 23, nil, nil ]\r\n```\r\nOn failure, the interaction looks like this:\r\n```\r\nSent: [ 0, 26, \"nvim_call_function\", [ \"rpcrequest\", [ 11, \"rpc-1\" ] ] ]\r\nSent: [ 0, 27, \"nvim_call_function\", [ \"rpcrequest\", [ 11, \"rpc-2\" ] ] ]\r\nReceived: [ 0, 1, \"rpc-1\", [] ]\r\nReceived: [ 0, 2, \"rpc-2\", [] ]\r\nSent: [ 1, 2, nil, nil ]\r\nSent: [ 1, 1, nil, nil ]\r\nException: (Failure \"The connection was closed!\").\r\n```\r\nWhen the failure happens, a log line appears in the low-level log file that says:\r\n```\r\nERROR {time} {nvim pid} call_set_error:615: RPC: ch {chan} returned a response with an unknown request id. Ensure the client is properly synchronized\r\n```\r\nIs this a bug, or am I doing something invalid? I'm replying to the requests in LIFO order to avoid running into https://github.com/neovim/neovim/issues/19932, and I'm not sure what else I could be doing that's invalid. I'm also curious whether this problem reproduces on 0.8.0, which I don't have available to me at the moment -- I see there have been changes to [parse_msgpack](https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/channel.c#L238) since the 0.7.0 release.",
    "closed_at": "2022-11-10T14:10:57Z",
    "comment_data": [
        {
            "author_association": "CONTRIBUTOR",
            "body": "The issue here seems to be that the plugin is sending the response to rpc-1 before Neovim has had an opportunity to reply to the plugin with the result of rpc-2.",
            "created_at": "2022-11-10T14:10:57Z",
            "html_url": "https://github.com/neovim/neovim/issues/20849#issuecomment-1310336063",
            "id": 1310336063,
            "issue_url": "https://api.github.com/repos/neovim/neovim/issues/20849",
            "node_id": "IC_kwDOAPphoM5OGiQ_",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/neovim/neovim/issues/comments/1310336063/reactions"
            },
            "updated_at": "2022-11-10T14:10:57Z",
            "url": "https://api.github.com/repos/neovim/neovim/issues/comments/1310336063",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/802511?v=4",
                "events_url": "https://api.github.com/users/ddickstein/events{/privacy}",
                "followers_url": "https://api.github.com/users/ddickstein/followers",
                "following_url": "https://api.github.com/users/ddickstein/following{/other_user}",
                "gists_url": "https://api.github.com/users/ddickstein/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/ddickstein",
                "id": 802511,
                "login": "ddickstein",
                "node_id": "MDQ6VXNlcjgwMjUxMQ==",
                "organizations_url": "https://api.github.com/users/ddickstein/orgs",
                "received_events_url": "https://api.github.com/users/ddickstein/received_events",
                "repos_url": "https://api.github.com/users/ddickstein/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/ddickstein/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/ddickstein/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/ddickstein"
            }
        }
    ],
    "comments": 1,
    "comments_url": "https://api.github.com/repos/neovim/neovim/issues/20849/comments",
    "created_at": "2022-10-28T15:34:00Z",
    "events_url": "https://api.github.com/repos/neovim/neovim/issues/20849/events",
    "html_url": "https://github.com/neovim/neovim/issues/20849",
    "id": 1427432943,
    "labels": [
        {
            "color": "c7def8",
            "default": false,
            "description": null,
            "id": 242522707,
            "name": "rpc",
            "node_id": "MDU6TGFiZWwyNDI1MjI3MDc=",
            "url": "https://api.github.com/repos/neovim/neovim/labels/rpc"
        }
    ],
    "labels_url": "https://api.github.com/repos/neovim/neovim/issues/20849/labels{/name}",
    "locked": false,
    "milestone": null,
    "node_id": "I_kwDOAPphoM5VFOXv",
    "number": 20849,
    "performed_via_github_app": null,
    "reactions": {
        "+1": 0,
        "-1": 0,
        "confused": 0,
        "eyes": 0,
        "heart": 0,
        "hooray": 0,
        "laugh": 0,
        "rocket": 0,
        "total_count": 0,
        "url": "https://api.github.com/repos/neovim/neovim/issues/20849/reactions"
    },
    "repository_url": "https://api.github.com/repos/neovim/neovim",
    "state": "closed",
    "state_reason": "completed",
    "timeline_url": "https://api.github.com/repos/neovim/neovim/issues/20849/timeline",
    "title": "connection closed due to \"unknown request id\" with non-deterministic behavior",
    "updated_at": "2023-07-20T13:33:27Z",
    "url": "https://api.github.com/repos/neovim/neovim/issues/20849",
    "user": {
        "avatar_url": "https://avatars.githubusercontent.com/u/802511?v=4",
        "events_url": "https://api.github.com/users/ddickstein/events{/privacy}",
        "followers_url": "https://api.github.com/users/ddickstein/followers",
        "following_url": "https://api.github.com/users/ddickstein/following{/other_user}",
        "gists_url": "https://api.github.com/users/ddickstein/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/ddickstein",
        "id": 802511,
        "login": "ddickstein",
        "node_id": "MDQ6VXNlcjgwMjUxMQ==",
        "organizations_url": "https://api.github.com/users/ddickstein/orgs",
        "received_events_url": "https://api.github.com/users/ddickstein/received_events",
        "repos_url": "https://api.github.com/users/ddickstein/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/ddickstein/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/ddickstein/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/ddickstein"
    }
}